### ISSUE TITLE
Activation functions: F.silu() returns a different result from x * torch.sigmoid(x)

### ISSUE THREAD

**Issue by testt-1** [pytorch/pytorch#88558]
### 🐛 Describe the bug

The silu/swish activation functions is defined as x * sigmoid(x). The implementation through the functional library (F.silu()) gives me different result than from the torch library -- written as x * sigmoid(x). The performance difference is significant, but the function that outperforms the other is not consistent; sometimes, F.silu() does better, and sometimes x * torch.sigmoid(x) does better. My model is seeded properly; i'm using the same seed when I run different activations. I made sure of this by running the model on the same activation function and getting the same result. Why is this happening? how is torch.sigmoid implemented? I'm trying to create my own activation functions, so this information is very important to me.

### Versions

Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: macOS 12.3 (arm64)
GCC version: Could not collect
Clang version: 13.1.6 (clang-1316.0.21.2.5)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 | packaged by conda-forge | (main, May 27 2022, 17:00:33)  [Clang 13.0.1 ] (64-bit runtime)
Python platform: macOS-12.3-arm64-arm-64bit
Is CUDA available: N/A
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: N/A

Versions of relevant libraries:
[pip3] No relevant packages
[conda] No relevant packages

cc @VitalyFedyunin @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

**Comment by jingxu10** [comment]
not reproducible. Any reproducer?
![image](https://user-images.githubusercontent.com/25193121/200196184-7b019161-ad35-4156-bbcf-e7f8101c3409.png)

**Comment by maryam-abdool** [comment]
> 

There are two issues with this this. First you're not using the functional library. It's supposed to be torch.nn.functional.silu(). Also, this is probably more of a numerical accuracy/stability issue that will not be shown unless you run the activation function through a deep learning model where values are updated. Doing simple prints on one tensor will give you equivalent results.

**Comment by maryam-abdool** [comment]
> not reproducible. Any reproducer?
> ![image](https://user-images.githubusercontent.com/25193121/200196184-7b019161-ad35-4156-bbcf-e7f8101c3409.png)
There are two issues with this this. First you're not using the functional library. It's supposed to be torch.nn.functional.silu(). Also, this is probably more of a numerical accuracy/stability issue that will not be shown unless you run the activation function through a deep learning model where values are updated. Doing simple prints on one tensor will give you equivalent results.

**Comment by jingxu10** [comment]
Does this happen on CPU or CUDA?

**Comment by maryam-abdool** [comment]
CPU.

**Comment by jingxu10** [comment]
Is there any reproducer available for investigation?

**Comment by maryam-abdool** [comment]
> Is there any reproducer available for investigation?

yes. I'm running the activation function on a model. I think it would be best if I upload the code on my GitHub and link it here. Is that fine?

**Comment by jingxu10** [comment]
yes, that sounds good.

**Comment by maryam-abdool** [comment]
I wrote simplified code to show the issue instead. Here it is. This will run it for x * torch.sigmoid(x). Comment     activation_function = SiLU() and uncomment out activation_function = Functional_SiLU() to run it for F.silu(). The code:

```
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Nov  6 17:26:38 2022

@author: mabbbs
"""

# Imports

# Import basic libraries
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
from collections import OrderedDict

# Import PyTorch
import torch # import main library
from torch.autograd import Variable
import torch.nn as nn # import modules
from torch.autograd import Function # import Function to create custom activations
from torch.nn.parameter import Parameter # import Parameter to create custom activations with learnable parameters
from torch import optim # import optimizers for demonstrations
import torch.nn.functional as F # import torch functions
from torchvision import datasets, transforms # import transformations to use for demo
import random

# helper function to train a model
def train_model(model, trainloader):
    '''
    Function trains the model and prints out the training log.
    INPUT:
        model - initialized PyTorch model ready for training.
        trainloader - PyTorch dataloader for training data.
    '''
    #setup training

    #define loss function
    criterion = nn.NLLLoss()
    #define learning rate
    learning_rate = 0.003
    #define number of epochs
    epochs = 5
    #initialize optimizer
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)

    #run training and print out the loss to make sure that we are actually fitting to the training set
    print('Training the model. Make sure that loss decreases after each epoch.\n')
    for e in range(epochs):
        running_loss = 0
        for images, labels in trainloader:
            images = images.view(images.shape[0], -1)
            log_ps = model(images)
            loss = criterion(log_ps, labels)

            optimizer.zero_grad()
            loss.backward()
            optimizer.step()

            running_loss += loss.item()
        else:
            # print out the loss to make sure it is decreasing
            print(f"Training loss: {running_loss}")

# simply define a silu function
def silu(input):
    '''
    Applies the Sigmoid Linear Unit (SiLU) function element-wise:
        SiLU(x) = x * sigmoid(x)
    '''
    return input * torch.sigmoid(input) # use torch.sigmoid to make sure that we created the most efficient implemetation based on builtin PyTorch functions

# create a class wrapper from PyTorch nn.Module, so
# the function now can be easily used in models
class SiLU(nn.Module):
    '''
    Applies the Sigmoid Linear Unit (SiLU) function element-wise:
        SiLU(x) = x * sigmoid(x)
    Shape:
        - Input: (N, *) where * means, any number of additional
          dimensions
        - Output: (N, *), same shape as the input
    References:
        -  Related paper:
        https://arxiv.org/pdf/1606.08415.pdf
    Examples:
        >>> m = silu()
        >>> input = torch.randn(2)
        >>> output = m(input)
    '''
    def __init__(self):
        '''
        Init method.
        '''
        super().__init__() # init the base class

    def forward(self, input):
        '''
        Forward pass of the function.
        '''
        return silu(input) # simply apply already implemented SiLU

# simply define a silu function
def functional_silu(input):
    '''
    Applies the Sigmoid Linear Unit (SiLU) function element-wise (functional)
    '''
    return F.silu(input) # use torch.sigmoid to make sure that we created the most efficient implemetation based on builtin PyTorch functions

class Functional_SiLU(nn.Module):
    '''
    Applies the Sigmoid Linear Unit (SiLU) function element-wise:
        SiLU(x) = x * sigmoid(x)
    Shape:
        - Input: (N, *) where * means, any number of additional
          dimensions
        - Output: (N, *), same shape as the input
    References:
        -  Related paper:
        https://arxiv.org/pdf/1606.08415.pdf
    Examples:
        >>> m = silu()
        >>> input = torch.randn(2)
        >>> output = m(input)
    '''
    def __init__(self):
        '''
        Init method.
        '''
        super().__init__() # init the base class

    def forward(self, input):
        '''
        Forward pass of the function.
        '''
        return functional_silu(input) # simply apply already implemented SiLU

def main():
    
    print('Loading the Fasion MNIST dataset.\n')
    random.seed(1)
    np.random.seed(1)
    torch.manual_seed(1)
    
    # Define a transform
    transform = transforms.Compose([transforms.ToTensor()])

    # Download and load the training data for Fashion MNIST
    trainset = datasets.FashionMNIST('~/.pytorch/F_MNIST_data/', download=True, train=True, transform=transform)
    trainloader = torch.utils.data.DataLoader(trainset, batch_size=64, shuffle=False)

    # 1. SiLU demonstration with model created with Sequential
    # use SiLU with model created with Sequential

    # initialize activation function
    activation_function = SiLU()
    #activation_function = Functional_SiLU()

    # Initialize the model using nn.Sequential
    model = nn.Sequential(OrderedDict([
                          ('fc1', nn.Linear(784, 256)),
                          ('activation1', activation_function), # use SiLU
                          ('fc2', nn.Linear(256, 128)),
                          ('bn2', nn.BatchNorm1d(num_features=128)),
                          ('activation2', activation_function), # use SiLU
                          ('dropout', nn.Dropout(0.3)),
                          ('fc3', nn.Linear(128, 64)),
                          ('bn3', nn.BatchNorm1d(num_features=64)),
                          ('activation3', activation_function), # use SiLU
                          ('logits', nn.Linear(64, 10)),
                          ('logsoftmax', nn.LogSoftmax(dim=1))]))

    # Run training
    print('Training model with SiLU activation.\n')
    train_model(model, trainloader)
    

if __name__ == '__main__':
    main()
```

**Result for x * torch.sigmoid(x):**
Loading the Fasion MNIST dataset.

Training model with SiLU activation.

Training the model. Make sure that loss decreases after each epoch.

Training loss: 462.2620523124933
Training loss: 345.3563243150711
Training loss: 309.7548027858138
Training loss: 284.4865501821041
Training loss: 269.7550822645426

**Result for F.silu(x):**
Loading the Fasion MNIST dataset.

Training model with SiLU activation.

Training the model. Make sure that loss decreases after each epoch.

Training loss: 462.26396933197975
Training loss: 345.43089351058006
Training loss: 309.6794162020087
Training loss: 285.47320745140314
Training loss: 269.3993896022439

you'll notice the difference starts out very small (after the second decimal point) and then becomes larger.

**Comment by jingxu10** [comment]
Thanks for sharing the reproducer. We will have a try.

**Comment by maryam-abdool** [comment]
@jingxu10 For reference, I've run this on an M1 Mac (arm), not Intel.

**Comment by albanD** [comment]
Also note that ~1e-6 differences between these functions is expected due to floating point standard specifications. And so it is expected that your training will not behave exactly the same.
If the training is stable, the small numerical difference shouldn't be a problem though.

**Comment by koya-ken** [comment]
This problem still seems to occur.
Is anyone else working on this problem?

You can check with the following code
Notice that the result of the subtraction is not zero because the difference is small.

```
import torch
import torch.nn as nn
import platform

def silu1(x):
    m = nn.SiLU()
    return m(x)

def silu2(x):
    sigmoid = nn.Sigmoid()
    return x * sigmoid(x)

def main():
    torch.random.manual_seed(0)
    pc = platform.uname()
    pc_os = pc.system
    pc_version = pc.version
    pc_arch = pc.machine
    print(f'torch version:{torch.__version__}')
    print(f'OS:{pc_os}, version:{pc_version}, arch:{pc_arch}')
    
    x = torch.randn(2)
    silu1_result = silu1(x)
    silu2_result = silu2(x)
    print(f'diff(cpu) silu1 - silu2={silu1_result - silu2_result}')

if __name__ == '__main__':
    main()
```

# result
```
torch version:2.0.1
OS:Windows, version:10.0.19045, arch:AMD64
diff(cpu) silu1 - silu2=tensor([-1.1921e-07,  0.0000e+00])
```

**Comment by albanD** [comment]
Hey!
As mentioned above, this is expected due to floating point precision: https://pytorch.org/docs/stable/notes/numerical_accuracy.html

**Comment by koya-ken** [comment]
I understand that there is a problem with floating point precision.
I am aware that if the calculations are the same, including order, the results will agree.

The fact that the calculation results do not match is detrimental when checking for errors in ONNX conversions, so

The issue has been closed, but I will leave the verification results for those who have the same problem.
It appears that pytorch silu does not use the sigmoid "function".
Therefore the order of calculation is different.
Note that silu is decomposed into sigmoid when converted to onnx.

The following code will match the output.

```
import torch
import torch.nn as nn
import platform

def silu1(x):
    m = nn.SiLU()
    return m(x)
def sigmoid(x):
    return 1 /  (1 + torch.exp(-x))
def silu2(x):
    # return x * sigmoid(x) # not match
    return x * 1 /  (1 + torch.exp(-x)) # match

def main():
    torch.random.manual_seed(0)
    pc = platform.uname()
    pc_os = pc.system
    pc_version = pc.version
    pc_arch = pc.machine
    print(f'torch version:{torch.__version__}')
    print(f'OS:{pc_os}, version:{pc_version}, arch:{pc_arch}')
    
    x = torch.randn(20)
    silu1_result = silu1(x)
    silu2_result = silu2(x)
    print(f'diff(cpu) silu1 - silu2={silu1_result - silu2_result}')

if __name__ == '__main__':
    main()
```
